pkg/sources/sourcer/sourcer.go

```diff
@@ -40,12 +39,9 @@ type SourceReader interface {
 	// idle watermarks should be published. Partition assignment to a pod is dynamic, so this method may return different
 	// partitions at different times. (Example - Kafka, every time topic rebalancing happens, the partitions gets updated)
 	Partitions(ctx context.Context) []int32
+	// Pending returns the pending messages number.
+	Pending(context.Context) (int64, error)
 }
 
-// Sourcer interface provides an isb.BufferReader abstraction over the underlying data source.
-// This is intended to be consumed by a connector like isb.forward
-type Sourcer interface {
-	SourceReader
-	forwarder.StarterStopper
```

Comment:
In this case, how to implement those logics for `preStop`, `preStart` or `postStart`? Like the `ready` flag used in HTTP source?

I understand for `preStart` or `postStart`, we could leverage the `New` function to sort of achieve (e.g. you start the kafka consumer Goroutine) that, however that is not an elegant implementation, and it relies on the `ctx` cancel to exit, and there's no way to do a clean shutdown.

Reply:
Synced up offline, we could go ahead.